### ISSUE TITLE
"Machine Learning - Introduction" Lecture Notes

### ISSUE THREAD

**Issue by hiyouga** [hiyouga/hiyouga-blog-project#42]
Experiment
Outcomes
Events
Probability

Review:
Random Variables
Definition
X: S to R (measurable)
measurable: Interval in a real line
#F should be all subsets of S
X is measurable with respect to F
Probability Calculations: PDF
P\_x(X) = likelihood of X = x
P[Event] = P[x\\in A]=\\int\_A P\_x(X)dx
Multiple RVs
High dimensional of random variables
Joint PDF: P[Event] = P[(X,Y)]\\in A]=\\iint\_A p\_{xy}(x,y)dxdy
Cond. PDF: P\_{x|y}[X|Y]=\\frac{p\_{xy}(X,Y)}{p\_y(Y)}

Statistics
E[X] = \\int\_\\infty xp\_x(X)dx
E[g(x)]=\\int\_\\infty g(x)p\_x(X)dx
E[g(x,y)]=\\iint g(x,y)p\_{xy}(X,Y)dxdy
E[X|Y]=\\int xP\_{(x|y)}(X|Y)dx

Classification Theory
X\\in R^n \\to y\\in\\{l\_1, \\cdots, l\_n\\}
Classical Classification Theory: optimal f() = f*()

Optimal Detection/Classification

Simple example: tumor detection(x-ray image)
x = intensity
x \\to f() \\to y\\in\\{H\_0,H\_1\\}
Question: What's the best f()?
Modeling the X or Y

Modeling:
y: P[H\_0] = P\_0
   P[H\_1] = P\_1

x: p\_{x|H\_0}(X|H\_0)=\\frac1{\\sqrt{2\\pi}\\sigma}e^{-(x-m\_0)/2\\simga^2}

P[H\_1|x] = \\frac{P[x|H\_1]*P\_1}{P\_x[X]}
f*(x)=H\_1, P[x|H\_1]*P\_1 > P[x|H\_0] * P\_0
=H\_0, P[x|H\_1]*P\_1 < P[x|H\_0] * P\_0

f*(): genrenal for n-dim x
optimal f() = f*()
k* = argmax\_{1<=k<=K}P[H\_k|x]
=argmax\_{1<=k<=K}P[x|H\_k]*P\_k

Bayes Classification

Gaussian PDF:
2D:
p(x)=\\frac1{\\sqrt{2\\pi}\\sigma}e^{-\\frac{(x-m)^2}{2\\sigma^2}}
X~N(m, \\sigma^2)

nD:
X=[x1, ..., xm]
X~N(m, \\Sigma)

Gaussian Q function:
p(x) = \\frac1{\\sqrt{2\\pi}}e^{-x^2/2}
Q(t) = \\int\_t^\\infty p(x)dx
Estimation/Regression

The Estimation Problem
Density Estimation(PDF): Data--p(X|H0) p(X|H1) Data--Decision boundary
Variable Estimation / Function Estimation: Conditional Expection:Data--E[X|Y]--g(Y)
Data--E[Y|X]--g(X)

Density Estimation Problem:
D:data X1,X2,...,Xn
xi~X xi is a sample variable  i.i.d. (indepandent and from same distribution)
D--P\_x(x)

parametric (e.g. Gaussian) 参数估计 parameter estimation
non. parametric 非参数估计 infinite parameters

bayesian (e.g. m of a Gaussian) has prior knowledge
frequentist 

four approaches

Parameter Estimation

Example: Gaussian
D = {x1, x2, ..., xn} i.i.d. ~ X~N(m, \\sigma^2)
(m, \\sigma^2) = \\theta = parameters

Estimation Problem:
D -- \\hat{\\theta}  the estimate of \\theta
Frequentist: Maximum likelihood estimation (MLE) 极大似然估计
\\hat{\\theta} = argmax\_\\theta p(x1, ..., xn | \\theta)
= argmax\_\\theta \\log o(x1, ..., xn | \\theta)  (log is monotonic function)

what is \\hat{m}

\log p(x1, .., xn | \\theta)
= log \\Pi p(xi | \\theta)
= \\Sigma \\log p(xi | \\theta)
f = \\Sigma(\\log\\frac1{\\sqrt{2\\pi}\\sigma}-\\frac{xi-m}{2\\sigma^2})

\\frac {\\partial f}{\\partial m}=0

\\Sigma(xi - m) = 0
\\hat{m}=\\frac1n\\Sigma xi

n -- \\infty
\\hat{\\theta} -- \\theta\* (ground truth)

Bayesian Estimation
Introduce prior p(\\theta): PDF for \\theta
Example: p(\\theta) \\theta = m, \\sigma^2 are known

Bayesian / MAP (Max a posteriori probability) estimation 最大后验估计
\\hat(\\theta) = argmax\_\\theta p(\\theta | x1, ..., xn)
= argmax \\log p(\\theta | x1, ..., xn)
= argmax \\log p(x1, ..., xn | \\theta) + \\log p(\\theta)  (Use Bayes)

for Gaussian
\\log p(x1, ..., xn | \\theta) + \\log p(\\theta)
= \\log \\Pi p(xi | \\theta) + \\log p(\\theta)
= f
\\frac{\\partial f}{\\partial m} = 0

\\hat{m} = \\frac{n / \\sigma^2}{n / \\sigma^2 + 1 / \\sigma\_0^2}(\\frac{\\Sigma xi}m) + \\frac{1 / \\sigma^2}{n / \\sigma^2 + 1 / \\sigma\_0^2}m\_0

Bayes Estimation is a linear combination of MLE and prior knowledge

if \\sigma^2 -- \\infty, the former goes to 0
means the data is noisy, it has large variance, and it is unreliable
\\hat{m} -- m\_0
if \\sigma\_0^2 -- \\infty, the latter goes to 0
means the prior is unreliable
\\hat{m} -- \\frac{\\Sigma xi}m

Variable / Function Estimation
Two random Variables X and Y
p(X, Y): joint PDF
-- p(Y | X)
-- p(X), p(Y)

Use X to estimate Y
\\hat{Y} = g(X) = estimation of Y using X
Question: what is the best g()
Prediction error = E[|Y - \\hat{Y}|] is non-differentiable
-- MSE (mean square error) = E[(Y - \\hat{Y})^2] = E[(Y - g(X))^2] = E[g]
\\frac{\\partial E}{\\partial g}=0

g(X) = aX + b linear estimation

E = E[(Y - \\hat{Y})^2]
= \\int\_\\infty dx\\int dy(y - \\hat{y})^2p\_{xy}(x, y)   (p\_{xy}(x, y) = p(y|x)p(x))
= \\int\_\\infty dx p(x) \\int\_\\infty dy(y - \\hat{y})^2 p(y | x)
min E with respect to \\hat{y}
for each x if we min \\int\_\\infty(y-\\hat{y})^2p(y|x)dy = Q
\\frac{\\partial Q}{\\partial y}=\\int\_\\infty -2(y-\\hat{y})p(y|x)dy=0
== \\int\_\\infty y p(y|x)dy = \\hat{y} \\int\_\\infty p(y|x)dy
== \\hat{y} = \\int\_\\infty y p(y|x)dy
\\hat{Y} = E[Y | X]

HW:
A: x = 0, y = 0, y = 1 - x
p(x, y) = 2, (x, y) \\in A; 0, otherwise
Find: \\hat{Y}=E[Y|X]=g\*(x)
Solution: E[Y|X] = \\int\_\\infty yP(y|x)dy
P(y|x) = \\frac{P(x,y)}{P(x)} = \\frac1{1-x}, (x,y)\\in A; 0, otherwise
P(x) = 2(1-x), x\\in[0,1]; 0, otherwise
E[Y|X] = \\int\_\\infty yP(y|x)dy = 0, y \\notin [0,1];
\\int\_0^{1-x} y\\frac1{1-x} dy = \\frac12(1-x) = g\*(X) = \\hat{Y}

\\hat{Y} = E[Y|X] = \\frac12(1-X) is the correct answer

Review: Classical Theory of Classification & Estimation
Classification:
Optimal Classifier Bayes/min Probability of Error
Two classes: calculate P[H\_1 | X] >&< P[H\_0 | X] (P[X | H\_1]P\_1 >&< P[X | H\_0]P\_0)

if multiple classes: (more than two classes)
classes: H\_k: k th classe, k = 1, 2, ..., k
H\_{k^/*} = argmax\_{H\_k} P[H\_k | X]

Estimation:
Density estimation: data -- P[X | H\_k] k = 0,1
Variable / Function estimation: X -- \\hat{Y}

- parametric: P[x|H\_k]=P[x|H\_k, \\theta]
- nonparametric

approach:
MLE: \\hat{\\theta} = argmax\_\\theta \\log p(x\_1, ..., x\_n | \\theta)   (data: i.i.d)
MAP: \\hat{\\theta} = argmax\_\\theta \\{\\log p(x\_1, ..., x\_n | \\theta) + \\log p(\\theta)\\}

X \\to \\hat{Y} (\\hat{Y} = g(X))
mean square error solution: E[(Y - \\hat{Y})^2]
\\hat{Y} = E[Y|X] = g\*(X)

Linear and Logistic Regression / Estimation
Linear Regression / Estimation:
The problem: X and Y: random variables/vectors
special case: X: random vector, Y: random scalar
Want: estimate Y from X
example: Y = price at t+1(future), X = prices at t, t-1, t-2, ..., t-m
optimal solution: \\hat{Y} = E[Y|X]
Problem: requires p(y|x\_1, ..., x\_{m+1})
alternative: \\hat{Y} = E[Y|X] \\~= w^TX + b = [w^T, b][x 1]^T angmauted
= w^T X

\\hat{Y} = E[Y|X] \\~= w^TX
Q: How to find the best w?
A: \\mathscr{E} = E[(Y - w^TX)^2]
w\* = argmin\_w E(w)
take partial derivatives
gradient
\\nabla\_w E = 0
E[(-2)(Y - w^TX)X] = 0
E[XY] - E[XX^T]w = 0
w\* = (E[XX^T])^{-1}E[XY]

find E[XX^T] and E[XY] in practice (from data)
Data = D = {(x1, y1), (x2, y2), ..., (xn, yn)}
E[XX^T] \\~= \\frac1n \\sum x\_ix\_i^T
E[XY] \\~= \\frac1n \\sum x\_iy\_i
w\* \\~= (\\frac1n \\sum x\_ix\_i^T)^{-1}(\\frac1n \\sum x\_iy\_i)
problem: dim(x) is too high

Gradient Descent:
Review: min f(x)
\\nabla\_x f(x) = 0
problem: \\nabla\_xf(x)=0 difficult to solve
One Solution: iterative solution
x\_{n+1} = x\_n + d\_n
Gradient Descent: d\_n = -\\epsilon\\nabla\_xf(x\_n)
\\epsilon: learning rate/step size: small, non-negative
d\_n : direction of the fastest decay

direction derivatives: \\frac{\\partial f}{\\partial \\vec{n}} = \\vec{n}\\cdot\\nabla f

Apply to linear regression:
Recall: E = E((Y - w^TX)^2)
\\nabla\_wE = E[(-2)(Y - w^TX)X]
G.D. w\_{n+1} = w\_n - \\epsilon\\nabla\_wE(w\_n)
w\_{n+1} = w\_n - \\epsilonE[(-2)(Y - w^TX)X]
\\~= w\_n - \\epsilon\\frac1n\\sum\_{i=1}^n(-2)(y\_i-w\_n^Tx\_i)x\_i

problem: m is too large, pick M in n data (M << n) (mini-batch)

Logistic Regression: (for classification)
Problem: Optimal Classification: two classes
P[H\_1 | X] >&< P[H\_0 | X]
Problem: P[H\_k | X] (k = 0, 1) is difficult to 
Solution: Logisitic Regression
P[H\_1 | X] \\~= \\frac1{1 + e^{-w^TX}}
\\frac1{1+e^{-t}} = p(t) = logit function
P[H\_0 | X] = 1 - P[H\_1 | X]
if P[H\_1 | X] = \\frac1{1 + e^{-w^TX}}
and P[H\_0 | X] = 1 - \\frac1{1 + e^{-w^TX}}
Decision boundary? -- P[H\_0 | X] = P[H\_1 | X] -- w^TX = 0 -- linear

Use an optimal linear classifier to approx classifier.

How to find w?
use \\frac1{1 + e^{-w^TX}} to approx P[H\_1 | X]
\\frac{e^{-w^TX}}{1 + e^{-w^TX}} to approx P[H\_0 | X]
Mean Square Error is not a good measure.
A better measure is Cross Entropy. (KL-distance)

Example: binary r.v.
X 0 1
P 1/2 1/2

X 0 1
P 1/4 3/4

Information Theory
Cross Entropy H(P, Q)
H(P, Q) = - P[X = 0] \\log Q[X = 0] - P[X = 1] \\log Q[X = 1] # number of bits need to code X assuming while Q is true
+ P[X = 0] \\log P[X = 0] + P[X = 1] \\log P[X = 1] # H(P) = entropy tfor P : number of bits need to code X assuming while P is true

P[H\_1 | X] = truth Q[H\_1 | X]
P[H\_0 | X] = truth Q[H\_0 | X]
they are approxmation

H(P, Q)(X) = cross entropy between P and Q
Criterion to optimize/minimize E\_x[H(P, Q)(X)] = E(w)

G.D. w\_{n+1} = w\_n - \\epsilon\\nabla\_w E(w\_n)
E[] \\~= \\frac1n E[]

—— Jun Zhang
